### DISCUSSION TITLE
FreshRSS and Google Reader API support now available

### DISCUSSION THREAD

**Discussion by jocmp** [jocmp/capyreader#492] [Announcements]
I’m happy to announce initial support for [FreshRSS](https://www.freshrss.org/) and the Google Reader API in the [latest dev update](https://github.com/jocmp/capyreader/releases/tag/2024.11.1071-dev). This was the [most requested service](https://github.com/jocmp/capyreader/discussions/281) from a poll I sent out several months ago. The integration offers the following features

- Fetch articles including favorites
- Mark an article as read or unread
- Mark an article as a favorite
- Add and edit feeds

This feature set is inline with the existing abilities offered for local and Feedin accounts in Capy Reader today.

I tested both FreshRSS and a separate instance of [Google Reader for Miniflux](https://miniflux.app/docs/google_reader.html). I’m sure I will have missed something about this integration, especially with other services that I don’t have access to like BazQux. Please let me know in this thread or file a [bug report](https://github.com/jocmp/capyreader/issues/new?labels=bug&template=bug_report.yml) if you run into any issues. I’ll follow up as soon as possible.

<img src="https://github.com/user-attachments/assets/06faaf77-47c1-4207-b846-d09152dd62a7" width="350px">

**Comment by Alkarex** [comment]
Excellent! Feel free to let me know if you need any sparing from FreshRSS side.
I am about to add CapyReader to our [list of supported clients](https://github.com/FreshRSS/FreshRSS/blob/edge/README.md#apis--native-apps).
Here are a few feature requests to improve the support:
* [ ] https://github.com/jocmp/capyreader/discussions/531
* [ ] https://github.com/jocmp/capyreader/discussions/532
* [ ] https://github.com/jocmp/capyreader/discussions/533

**Reply by jocmp** [reply]
Hey @Alkarex thanks for reaching out! Your docs on the API were super helpful when implementing this. 

I'll take a look at the tickets you linked. I definitely want to add labels at some point soon.

**Comment by jstride** [comment]
I've been using this with fresh RSS and it's my GOTO app. Both great products and now with the mark as read on scroll it's amazing.

Thank you.

**Comment by st123m123** [comment]
Hi, I'm running a FreshRSS instance locally on my homeserver. When trying to connect via http I get the following message "CLEARTEXT communication is not permitted by network security policy", I guess as I do not connect via https?

With different RSS readers (e.g. FeedMe, Read You) there is no issue connecting through http. Is there a way I could connect via http?

**Reply by jocmp** [reply]
@st123m123 which version of the app are you running? There was a bug (https://github.com/jocmp/capyreader/commit/f08dfbca96c9d451eaba8a35741fc29c1423747f) in the network policy preventing HTTP/cleartext that should have been fixed as of [2025.01.1089](https://github.com/jocmp/capyreader/releases/tag/2025.01.1089).

**Reply by st123m123** [reply]
I was still using the previous version 2024.12.1087. Now updating to the latest version, it works. Many thanks

**Comment by montechristos** [comment]
I also set up a local freshrss instance.
Now trying to connect to it from Capy.
Am I missing something? I cannot seem to find the above screen for adding my freshrss?
I also tried directly with adding a feed and http://192.168.1.203/i/?c=feed&a=add&url_rss=%s but it does not work
I am running version 2025.02.1105

(any instructions would be appreciated)

**Reply by Alkarex** [reply]
@montechristos Make sure you are using the latest release of Capy Reader https://github.com/jocmp/capyreader/releases

**Reply by montechristos** [reply]
According to the releases, I should have this feature: 2025.02.1105

**Reply by Alkarex** [reply]
Did you check under settings / account?

**Reply by montechristos** [reply]
![account_screen](https://github.com/user-attachments/assets/185947e0-5073-4e8e-8c65-bb3cf8b7798d)
![version_screen](https://github.com/user-attachments/assets/c3a0fb3e-e4e1-4c2e-a3a3-56febd2820ae)
Yes, this is what I have

**Reply by jocmp** [reply]
@montechristos it looks like you clicked into the "Local" account instead of FreshRSS. Local in this case referring to the device, not the network. 

From the account settings screen, click "Delete Account" which will only wipe device local data.

After that, you'll be brought to a page where you should select "FreshRSS":

<img src="https://github.com/user-attachments/assets/550613c3-6348-4452-b86c-336832247974" width="400px">

**Reply by montechristos** [reply]
Ok, thanks. I thought I could just ‘add’ an account rather than replacing the existing one

It does not work yet though:
I put this address: http://freshrss.lan/api/greader.php/ with my username and password, but it fails:
Failed with status 503 Service Unavailable 
Any ideas?

**Reply by jocmp** [reply]
I haven't seen that before with FreshRSS but general, the app connected to a server, but the server failed. The exact error might show up in the FreshRSS server's logs ([docs](https://freshrss.github.io/FreshRSS/en/admins/logs_and_errors.html)).

**Reply by montechristos** [reply]
Solved: I had not enabled API on my freshrss installation. Also I was trying with my username/password, whereas I should have used the API password.
Thanks for your support @jocmp

**Reply by jocmp** [reply]
Sweet. Thanks for posting the solution too!

**Reply by l3pr-org** [reply]
> Solved: I had not enabled API on my freshrss installation. Also I was trying with my username/password, whereas I should have used the API password. Thanks for your support @jocmp

Thanks montechristos, this fixed my issue as well!

@jocmp is it possible to mark montechristos's response as the answer for better visibility? Moving to another discussion may be a better option as I see this discussion was originally intended as a feature announcement.

**Comment by margaretjoanmiller** [comment]
I'm writing my own RSS sync server and I'm ripping my hair out implementing the google reader api haha. I can see that the app gets a list of subscriptions but it never displays them or pulls articles from them. Could you provide the JSON capy expects for this list (and for the items endpoint too)? Or give some pointers? Thanks

**Reply by margaretjoanmiller** [reply]
This api does (mostly) work in newsboat so there's a subtle difference in the way capy understands the api that's tripping me up

**Reply by jocmp** [reply]
Hey @margaretjoanmiller, let me see if I can help. Capy fetches subscriptions and then adapts them ([source code](https://github.com/jocmp/capyreader/blob/8922f4428db39ac22f66a0b1e50ca3b5e6639da6/capy/src/main/java/com/jocmp/capy/accounts/reader/ReaderAccountDelegate.kt#L229-L231)) into a data model called a "feed" internally. The access is somewhat indirect so that all sources have a shared data model.

In terms of JSON, I think the most robust Google Reader documentation I found - and I used for Capy Reader - was the FeedHQ docs (https://feedhq.readthedocs.io/en/latest/api/reference.html). They also include an example with the subscription JSON format ([link](https://feedhq.readthedocs.io/en/latest/api/reference.html#subscription-list)):

```jsonc
{
    "subscriptions": [
        {
            "title": "A List Apart", // Capy uses this for the feed title
            "firstitemmsec": "1373999174000",
            "htmlUrl": "http://alistapart.com",
            "sortid": "B0000000",
            "id": "feed/http://feeds.feedburner.com/alistapart/main",  // Capy uses this to refresh feed entries
            "categories": [
                {
                    "id": "user/1/label/Tech",
                    "label": "Tech" 
                }
            ]
        }
    ]
}
```

FreshRSS also has a list of other resources for Google Reader API that may be useful: https://freshrss.github.io/FreshRSS/en/developers/06_GoogleReader_API.html. Let me know if that answers your questions!

**Reply by margaretjoanmiller** [reply]
Hmm I think I'm implementing that properly but still no luck. I'll do some digging in the source code to see if I can find anything.

**Comment by bonaldi** [comment]
Seems to be silently failing with Bazqux: I appear to login OK but the feeds never load.

**Reply by jocmp** [reply]
Hey @bonaldi, I never did get Bazqux working the first time. Let me take another look.
